README
What is OTRS?
=============

  OTRS is an Open source Ticket Request System with many features to manage
  customer telephone calls and e-mails. It is distributed under the GNU
  AFFERO General Public License (AGPL) and tested on Linux, Solaris, AIX,
  Windows, FreeBSD, OpenBSD and Mac OS 10.x. Do you receive many e-mails and
  want to answer them with a team of agents? You're going to love OTRS!

  You can find a list of features at:
  http://doc.otrs.org/3.1/en/html/features-of-otrs.html


License:
========
  It is distributed under the AFFERO GNU General Public License - see the
  accompanying COPYING file for more details.

Documentation:
==============
  You can find quick documentation in README.* and the long version online at
  http://doc.otrs.org/.

Software requirements:
======================

  Perl
    * Perl 5.8.8 or higher

  Webserver
    * Webserver with CGI support (CGI is not recommended)
    * Apache2+mod_perl2 or higher (recommended, mod_perl is really fast!)
    * IIS6 or higher

  Databases
    * MySQL 4.1 or higher
    * PostgreSQL 8.0 or higher
    * Oracle 10g or higher
    * Microsoft SQL Server 2005 or higher

Directories & Files:
====================
$HOME (e. g. /opt/otrs/)
|
|  (all executables)
|--/bin/             (all system programs)
|   |--/otrs.PostMaster.pl      (email2db)
|   |--/otrs.PostMasterMail.pl  (email2db)
|   |--/cgi-bin/
|   |   |----- /index.pl        (Global Agent & Admin handle)
|   |   |----- /customer.pl     (Global Customer handle)
|   |   |----- /public.pl       (Global Public handle)
|   |   |----- /installer.pl    (Global Installer handle)
|   |   |----- /nph-genericinterface.pl (Global GenericInterface handle)
|   |--/fcgi-bin/               (If you're using FastCGI)
|
|  (all modules)
|--/Kernel/
|   |-----/Config.pm      (main configuration file)
|   |---- /Config/        (Configuration files)
|   |      |---- /Files/  (System generated, don't touch...)
|   |
|   |---- /Output/        (all output generating modules)
|   |      |---- /HTML/
|   |             |--- /Standard/*.dtl (all dtl files for Standard theme)
|   |
|   |--- /GenericInterface (GenericInterface framework)
|          |--- /Invoker/ (invoker backends)
|          |--- /Mapping/ (data mapping backends)
|          |--- /Operation/ (operation backends)
|          |--- /Transport/ (network transport backends)
|   |
|   |---- /Language/      (all translation files)
|   |
|   |---- /Modules/        (all action modules e. g. QueueView, Move, ...)
|   |      |----- /Admin*      (all modules for the admin interface)
|   |      |----- /Agent*      (all modules for the agent interface)
|   |      |----- /Customer*   (all modules for the customer interface)
|   |
|   |---- /System/         (back-end API modules, selection below)
|           |--- /Auth.pm        (authentication module)
|           |--- /AuthSession.pm (session authentication module)
|           |--- /DB.pm          (central DB interface)
|           |--- /DB/*.pm        (DB drivers)
|           |--- /DynamicField.pm (Interface to the DynamicField configuration)
|           |--- /DynamicField
|                 |--- /Backend.pm (Interface for using the dynamic fields)
|                 |--- /Backend/*.pm (DynamicField backend implementations)
|                 |--- /ObjectType/*.pm (DynamicField object type implementations)
|           |--- /Email.pm       (create and send e-mail)
|           |--- /EmailParser.pm (parsing e-mail)
|           |--- /GenericInterface/*.pm (all DB related GenericInterface modules)
|           |--- /Group.pm       (group module)
|           |--- /Log.pm         (log module)
|           |--- /Queue.pm       (information about queues. e. g. responses, ...)
|           |--- /Scheduler      (Scheduler files)
|                 |--- /TaskHandler/ (task handler backends for the Scheduler)
|           |--- /Ticket.pm      (ticket and article functions)
|           |--- /User.pm        (user module)
|           |--- /Request.pm    (HTTP/CGI abstraction module)
|
|  (data stuff)
|--/var/
|   |--/article/               (all incoming e-mails, plain 1/1 and all attachments ...
|   |                            ... separately (different files), if you want to store on disk)
|   |--/cron/                  (all cron jobs for escalations and such)
|   |
|   |--/fonts/                 (true type fonts for PDF generation)
|   |
|   |--/httpd/                 (all static files served by HTTP)
|   |   |--- /htdocs/
|   |         |--- /js/        (javascript files for OTRS)
|   |               |--- /js-cache/        (auto-generated minified JS files)
|   |               |--- /thirdparty/      (contains jQuery, CKEditor and other external JS libraries)
|   |         |--- /skins/     (CSS and images for front end)
|   |               |--- /Agent/        (Agent skins)
|   |                     |--- /default/ (default skin)
|   |                           |--- /css/ (stylesheets)
|   |                           |--- /css-cache/ (auto-generated minified CSS files)
|   |                           |--- /img/ (images)
|   |                     |--- /slim/    (additional skin)
|   |                           |--- /.../ (files)
|   |                     |--- /ivory/   (additional skin)
|   |                           |--- /.../ (files)
|   |               |--- /Customer/     (Customer skins)
|   |                     |--- /default/ (default skin)
|   |                           |--- /.../ (files)
|   |                     |--- /ivory/
|   |                           |--- /.../ (files)
|   |
|   |--/log/                   (log files)
|   |   |--/TicketCounter.log  (ticket counter)
|   |
|   |--/sessions/              (session info)
|   |
|   |--/spool/                 (spool files)
|   |
|   |--/stats/                 (statistics)
|   |
|   |--/tmp/                   (temporary files, such as cache)
|
|  (tools stuff)
|--/scripts/
|   |----  /database/
|           |--- /otrs-schema.(mysql|postgresql|*).sql (create database script)
|           |--- /otrs-initial_insert.(mysql|postgresql|*).sql (all initial sql data - e. g.
|           |                                                   root user, queues, ...)
|           |--- /otrs-schema-post.(mysql|postgresql|*).sql (create foreign keys script)
|


CHANGES
# --
# CHANGES - change log of OTRS
# Copyright (C) 2001-2013 OTRS AG, http://otrs.org/
# --
# $Id: CHANGES,v 1.2260.2.206 2013-02-14 13:31:52 mg Exp $
# --
# This software comes with ABSOLUTELY NO WARRANTY. For details, see
# the enclosed file COPYING for license information (AGPL). If you
# did not receive this file, see http://www.gnu.org/licenses/agpl.txt.
# --

3.1.22 2015-??-??
- 2015-01-05 Enhanced Permission Checks in GenericInterface Ticket Connector.

3.1.21 2014-04-01
 - 2014-03-18 Fixed bug#10374 - OTRS does not protect against clickjacking.
 - 2014-03-13 Fixed bug#10361 - Incorrect handling of special characters in DynamicFields.
 - 2014-02-27 Fixed bug#10309 - mails with an empty return-path header must not trigger auto responses in OTRS.

3.1.20 2014-02-25
 - 2014-02-21 Fixed bug#10287 - AdminEmail does not work / causes internal server error.
 - 2014-02-20 Improved HTML filter.

3.1.19 2014-01-28
 - 2014-01-14 Fixed bug#10158 - Missing quoting in State::StateGetStatesByType().
 - 2013-12-20 Fixed bug#10099 - Missing challenge token checks on customer interface.
 - 2013-09-20 Fixed bug#8489 - setting Tickets per page resets AgentTicketQueue.
 - 2013-09-06 Fixed bug#9661 - Useless code in DynamicField backend.
 - 2013-07-24 Fixed bug#9622 - Actions in Small ticket overview don't work when cookies are turned off.
 - 2013-07-22 Fixed bug#9541 - Package manager cannot use https proxy.
 - 2013-07-17 Fixed bug#9594 - No auto-reply sent with multiple From addresses in AgentTicketPhone on PostgreSQL and Oracle.
 - 2013-07-12 Fixed bug#3434 - Validity of search time frame not checked by OTRS.
 - 2013-07-10 Fixed bug#9596 - On merge and bounce screens is confusing when fill or not 'To', 'Subject' and 'Body' fields.
 - 2013-07-10 Fixed bug#9595 - Incomplete page reload handling in merge and bounce.
 - 2013-07-09 Fixed bug#3007 - CheckMXRecord and CheckEmailAddresses have no effect on AgentTicketBounce.
 - 2013-07-09 Fixed bug#9512 - Database error for invalid date in AgentTicketSearch.
 - 2013-07-09 Fixed bug#8835 - No article found for TicketID <TICKET ID> when showing group tickets
 - 2013-07-08 Fixed bug#9583 - Dynamic Fields of type Date have timestamp in notifications.
 - 2013-07-08 Fixed bug#9579 - SOAP Serializer used in Kernel/GenericInterface/Transport/HTTP/SOAP.pm does not correctly set namespace.
 - 2013-07-08 Fixed bug#7359 - Setting pending states via generic agent does not set pending time.
 - 2013-07-08 Fixed bug#8380 - Middle name not displayed in AdminCustomerUser.
 - 2013-07-05 Fixed bug#9576 - GI TicketSearch Date and Date/Time dynamic fields are ignored.
 - 2013-07-04 Changed Dynamic Field SearchFieldParameterBuild() API, LayoutObject is now optional.
 - 2013-07-04 Fixed bug#9573 - Date and DateTime dynamic fields not considered in GenericAgent Jobs.

3.1.18 2013-07-09
 - 2013-07-03 Fixed bug#9561 - ACL restriction with CustomerID for DynamicFields at new Ticket screen not working.
 - 2013-07-02 Fixed bug#9425 - Wrong created date for queue view.
 - 2013-06-28 Fixed bug#9125 - AgentTicketSearch dialog does not expand when choosing more search criteria.
 - 2013-06-28 Fixed bug#8273 - Copying text in preview mode not possible.
 - 2013-06-28 Fixed bug#9557 - Cannot see quoted text in customer ticket zoom.
 - 2013-06-27 Fixed bug#9011 - GenericInterface: New value after value mapping can't be 0.
 - 2013-06-25 Improved parameter quoting in various places.
 - 2013-06-24 Fixed bug#9104 - Group permission for customer subset overwrites permissions for other customers.
 - 2013-06-13 Fixed bug#8719 - PasswordMin2Lower2UpperCharacters problem.

3.1.17 2013-06-18
 - 2013-06-06 Improved permission checks in AgentTicketWatcher.
 - 2013-06-06 Fixed bug#9503 - no connection header in soap responses.
 - 2013-06-04 Added parameter "-t dbonly" to backup.pl to only backup the database
 - 2013-05-31 Fixed bug#9491 - GenericAgent job update with dynamic fields sends Uninitialized value error.
 - 2013-05-24 Fixed bug#9462 - Package Management page timeout due to HTTPS disabled on Proxy connections.

3.1.16 2013-05-21
 - 2013-05-17 Updated Package Manager, that will ensure that packages to be installed
    meet the quality standards of OTRS Group. This is to guarantee that your package
    wasn’t modified, which may possibly harm your system or have an influence on the
    stability and performance of it. All independent package contributors will have
    to conduct a check of their Add-Ons by OTRS Group in order to take full advantage
    of the OTRS package verification.
 - 2013-05-16 Fixed bug#9387 - Error in a condition with dynamic fields in NotificationEvent.
 - 2013-05-14 Fixed bug#9286 - Ticket::ChangeOwnerToEveryone isn't functional, After a AJAX Load the setting is ignored.
 - 2013-05-14 Fixed bug#7518 - Escalation Notify by not working properly (follow-up fix).
 - 2013-05-07 Fixed bug#7478 - Got an external answer to an internal mail.
 - 2013-05-07 Improved permission checks in AgentTicketPhone.
 - 2013-05-02 Fixed bug#9360 - DynamicField Names shown in CSV output.
 - 2013-04-30 Fixed bug#9384 - Problem with Method ServiceParentsGet of ServiceObject.
 - 2013-04-23 Fixed bug#9072 - Reply to email-internal includes customer users email in Cc. field.

3.1.15 2013-04-23
 - 2013-04-15 Added Malay translation.
 - 2013-04-12 Fixed bug#8960 - AgentTicketSearch.pm SearchProfile problem.
 - 2013-04-09 Fixed bug#9182 - Customer Search Function -> If you go into a ticket and go back you got not the search results.
 - 2013-04-02 Fixed bug#9198 - Linked search with fulltext AND additional attributes.
 - 2013-03-27 Fixed bug#9295 - Article dynamic field is not searchable.
 - 2013-04-08 Fixed bug#9312 - LinkObject permission check problem.

3.1.14 2013-04-02
 - 2013-03-21 Fixed bug#9257 - No notifications to agents with out-of-office set but period not reached.
 - 2013-03-18 Improved permission checks in LinkObject.

scripts/webform.pl
#!/usr/bin/perl -w
# --
# webform.pl - a simple web form script to generate email with
# X-OTRS-Queue header for an OTRS system (x-headers for dispatching!).
# Copyright (C) 2001-2014 OTRS AG, http://otrs.com/
# --
# $Id: webform.pl,v 1.14 2009-02-20 12:05:54 mh Exp $
# --
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU AFFERO General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
# or see http://www.gnu.org/licenses/agpl.txt.
# --

use strict;
use warnings;

# to get the errors on screen
use CGI::Carp qw(fatalsToBrowser);

# Simple Common Gateway Interface Class
use CGI;

my $VERSION = qw($Revision: 1.14 $) [1];

# --------------------------
# web form options
# --------------------------
my $Ident = 'ahfiw2Fw32r230dddl2foeo3r';

# sendmail location and options
my $Sendmail = '/usr/sbin/sendmail -t -i -f';

# email where the emails of the form will send to
my $OTRSEmail = 'otrs-system@example.com';

# topics and dest. queues
my %Topics = (

    # topic => OTRS queue
    'Info'      => 'info',
    'Support'   => 'support',
    'Bugs'      => 'bugs',
    'Sales'     => 'sales',
    'Billing'   => 'billing',
    'Webmaster' => 'webmaster',
);

# --------------------------
# html header
# --------------------------

sub Header {
    my (%Param) = @_;

    # html quote
    for my $Value (qw(Title)) {
        $Param{$Value} = _Ascii2Html( $Param{$Value} );
    }

    my $Output = <<EOF;
Content-Type: text/html

<html>
<head>
    <title>$Param{Title}</title>
</head>
<body>

<h1>$Param{Title}</h1>
<hr>
EOF
    return $Output;
}

# -------------------------
# html footer
# -------------------------

sub Footer {
    my $Output = <<EOF;
<hr>

</body>
</html>
EOF
    return $Output;
}

# -------------------------
# Thanks
# -------------------------

sub Thanks {
    my (%Param) = @_;

    # html quote
    for my $Value (qw(From)) {
        $Param{$Value} = _Ascii2Html( $Param{$Value} );
    }

    my $Output = <<EOF;
Thanks <b>$Param{From}</b>! Your request is forwarded to us. <br>
We will answer ASAP.<br>
EOF
    return $Output;
}

# ----------------------
# error
# ----------------------

sub Error {
    my (%Param) = @_;

    # html quote
    for my $Value (qw(Message)) {
        $Param{$Value} = _Ascii2Html( $Param{$Value} );
    }

    my $Output = <<EOF;
<font color="red">$Param{Message}</font><br>
EOF
    return $Output;
}

# ------------------------
# start the real actions
# ------------------------
my $CGI      = new CGI;
my %GetParam = ();
for (qw(Action From FromEmail Subject Topic Body)) {
    $GetParam{$_} = $CGI->param($_) || '';
}

# what should I do?
if ( $GetParam{Action} eq 'SendMail' ) {
    SendMail(%GetParam);
}
else {
    WebForm();
}

# ------------------------
# web form
# ------------------------
sub WebForm {
    print Header( Title => 'Submit Request' );
    print '
    <form action="webform.pl" method="post">
        <input type="hidden" name="Action" value="SendMail">
        <table>
            <tr>
                <td>Topic:</td>
                <td>
';
    for my $Key ( sort keys %Topics ) {
        my $HTMLKey   = _Ascii2Html($Key);
        my $HTMLValue = _Ascii2Html( $Topics{$Key} );
        print $HTMLKey . '<input type="radio" name="Topic" value="' . $HTMLValue . '">';
    }
    print '
                </td>
            </tr>
            <tr>
                <td>From:</td>
                <td><input type="text" name="From" size="45" value=""></td>
            </tr>
            <tr>
                <td>Email:</td>
                <td><input type="text" name="FromEmail" size="45" value=""></td>
            </tr>
            <tr>
                <td>Subject:</td>
                <td><input type="text" name="Subject" size="45" value=""></td>
            </tr>
            <tr valign="top">
                <td>Message:</td>
                <td><textarea name="Body" rows="15" cols="45"></textarea></td>
            </tr>
            <tr>
                <td></td>
                <td><input type="submit" value="Submit"></td>
            </tr>
        </table>
    </form>
';
    print Footer();
}

# --------------------------
# send email
# --------------------------

sub SendMail {
    my (%Param) = @_;

    # check needed params
    my $Output = '';
    for (qw(From FromEmail Subject Topic Body)) {
        if ( !$Param{$_} ) {
            $Output .= Error( Message => "Param $_ is needed!" );
        }
    }
    if ($Output) {
        print Header( Title => 'Error!' ) . $Output . Footer();
        return;
    }

    # simple email check
    my $NonAscii     = "\x80-\xff";                            # Non-ASCII-Chars are not allowed
    my $Nqtext       = "[^\\\\$NonAscii\015\012\"]";
    my $Qchar        = "\\\\[^$NonAscii]";
    my $Protocol     = '(?:mailto:)';
    my $NormUser     = '[a-zA-Z0-9][a-zA-Z0-9_.-]*';
    my $QuotedString = "\"(?:$Nqtext|$Qchar)+\"";
    my $UserPart     = "(?:$NormUser|$QuotedString)";
    my $DomMainPart  = '[a-zA-Z0-9][a-zA-Z0-9._-]*\\.';
    my $DomSubPart   = '(?:[a-zA-Z0-9][a-zA-Z0-9._-]*\\.)*';
    my $DomTldPart   = '[a-zA-Z]{2,5}';
    my $DomainPart   = "$DomSubPart$DomMainPart$DomTldPart";
    my $Regex        = "$Protocol?$UserPart\@$DomainPart";

    if ( $Param{FromEmail} !~ /^$Regex$/ ) {
        my $Output = Header( Title => 'Error!' );
        $Output .= Error( Message => "Your email '$Param{FromEmail}' is invalid!" );
        $Output .= Footer();
        print $Output;
        return;
    }

    # quoting (take care to not injection any other header)
    for my $Key (qw(From FromEmail Topic Subject Topic)) {
        $Param{$Key} =~ s/(\n|\r)//g;
    }

    # build email
    my @Mail = ("From: $Param{From} <$Param{FromEmail}>\n");
    push @Mail, "To: $Param{Topic} <$OTRSEmail>\n";
    push @Mail, "Subject: $Param{Subject}\n";
    push @Mail, "X-OTRS-Ident: $Ident\n";
    push @Mail, "X-OTRS-Queue: $Param{Topic}\n";
    push @Mail, "X-OTRS-ArticleKey1: Sent via\n";
    push @Mail, "X-OTRS-ArticleValue1: Webform\n";
    push @Mail, "X-OTRS-ArticleKey2: Orig. sort\n";
    push @Mail, "X-OTRS-ArticleValue2: $Param{Topic}\n";
    push @Mail, "X-Mailer: OTRS WebForm ($VERSION)\n";
    push @Mail, "X-Powered-By: OTRS (http://otrs.org/)\n";
    push @Mail, "\n";
    push @Mail, $Param{Body};
    push @Mail, "\n";

    # send mail
    my $FromEmail = $Param{FromEmail};
    $FromEmail =~ s/"|;|'|<|>|\||\s|\r|\n|\t|`//ig;
    $FromEmail = quotemeta $FromEmail;
    if ( open my $Mail, '|-', "$Sendmail $FromEmail" ) {
        print $Mail @Mail;
        close $Mail;

        # thanks!
        my $Output = Header( Title => 'Thanks!' );
        $Output .= Thanks(%Param);
        $Output .= Footer();
        print $Output;
    }
    else {

        # error
        my $Output = Header( Title => 'Error!' );
        $Output .= Error( Message => "Can't send email: $!" );
        $Output .= Footer();
        print $Output;
    }
}

sub _Ascii2Html {
    my $Text = shift;

    $Text =~ s/&/&amp;/g;
    $Text =~ s/</&lt;/g;
    $Text =~ s/>/&gt;/g;
    $Text =~ s/"/&quot;/g;

    return $Text;
}
